docs/examples/Fine-Tuning-Function-Calling.md

```diff
@@ -0,0 +1,273 @@
+
+Levanter provides good support for both methods. Therefore, we can easily try both methods and compare their results.
+
+#### Full-weight Fine-tuning
+
+We start with full-weight fine-tuning. Below is our configuration. Noteably:
+
+- The base model is `meta-llama/Llama-2-7b-hf`. It is set as the default value, so we don't need to specify it explicitly.
+- Batch size: We set the batch size to 128, which is the maximum batch size that can fit into a single TPUv3-8.
+- Learning rate: We compared the results with 3 epochs vs 2 epochs, and found that 2 epochs is sufficient to achieve the best results, while 3 epochs leads to overfitting.
+
+```yaml
+data: "gs://levanter-data/fine-tuning/gem-viggo/GEM_viggo_train.jsonl"
+data_cache_dir: "gs://levanter-data/tokenized/GEM_viggo_llama/"
+trainer:
+  wandb:
+    project: "levanter"
+    tags: ["viggo", "llama", "full-weight"]
+  mp: p=f32,c=bfloat16
+  train_batch_size: 128
+  num_train_steps: 80  # 5103 examples / 128 batch size * 2 epochs
+  tensor_parallel_axes: ["mlp", "heads"]
+optimizer:
+  learning_rate: 2E-5
+```
+
+The detailed instruction to run the training job can be found in the [Fine-Tuning documentation](../Fine-Tuning.md).
+Here is the command to run the training job on TPU:
+
+```bash
+gcloud compute tpus tpu-vm ssh finetune-32 --zone us-east1-d --worker=all \
+--command="WANDB_API_KEY=${YOUR WANDB TOKEN HERE} \
+HUGGING_FACE_HUB_TOKEN=${YOUR HF TOKEN HERE} \
+bash levanter/infra/run.sh python \
+levanter/examples/alpaca/alpaca.py \
+--config_path gs://<config-yaml-file> \
+--hf_save_path gs://<somewhere>"
+```
+
+Given the small dataset and high efficiency of Levanter, the entire training job completed quickly in only 21 min on a single TPUv3-8.
+
+Below shows an example where we compare the output from the original chat model and the fine-tuned model. We can see that the fine-tuned model is able to generate the correct function call and precisely capture the right attributes, while the original model is not able to do so.
+
+```
+Query: SpellForce 3 is a pretty bad game. The developer Grimlore Games is clearly a bunch of no-talent hacks, and 2017 was a terrible year for games anyway.
+Target: give_opinion(name[SpellForce 3], release_year[2017], developer[Grimlore Games], rating[poor])
+Generation from Llama2-7B Chat: \nOutput: inform(name[SpellForce 3], release_year[2017], esrb[NA], genres[fantasy,\naction, adventure], platforms[PC], available_on_steam[no], has_linux_release[no],\nhas_mac_release[no])
+Generation from fine-tuned Llama2-7B: give_opinion(name[SpellForce 3], release_year[2017], developer[Grimlore Games], rating[poor])
+```
+
+#### LoRA Fine-tuning
+
+Below is our configuration for LoRA fine-tuning. Note that it is very similar to the full-weight fine-tuning configuration, except for a few differences:
+
+- We increased the number of steps by 1 more epoch. LoRA fine-tuning is more efficient and less prone to overfitting, so we can train for more steps.
```

Comment:
more efficient should be qualified a bit. My experience is it's slower but more ram and parameter efficient. lower capacity means it's regularized better